### PR TITLE
Initial support for Pipeline Parallelism

### PR DESCRIPTION
**Pull request by michaelbenayoun**
This PR adds support for Pipeline Parallelism on a **single node** for the **Llama** architecture.
Multi-node training and other possible relevant architectures will be added in later PRs.

### PP works

- [x] Model can be parallelized
- [x] Lazy loading works perfectly with PP
- [x] The optimizer is created properly
- [x] The optimizer step works properly
- [x] The logging of the loss works
- [x] Model can train
- [ ] Evaluation works

### Variants

Only tested via tests here (it will be tested more in the PR for multi-node training).

- [x] TP + PP works
- [x] DP + PP works (without ZeRO-1)
- [x] DP + PP works (with ZeRO-1)

## Checkpointing

- [x] Can save sharded checkpoints
- [x] Can resume from sharded checkpoint
- [x] Update the consolidation function / command ?

## Tests

- [x] Tests in `tests/distributed/test_model_parallelization.py`
- [x] Tests in `tests/test_examples.py`
- [x] New extensive test suite `tests/distributed/test_common.py`

## Other

- [x] The cache system can download and upload compilation files from a training with PP
- [x] Update the examples

### CODE REVIEW

File: optimum/neuron/distributed/utils.py [hunk 1]
```diff
@@ -753,6 +841,21 @@ def make_optimizer_constructor_lazy(optimizer_cls: Type["torch.optim.Optimizer"]
 
     def optimizer_constructor(*args, **kwargs):
         optimizer_with_no_parameters = optimizer_cls([torch.nn.Parameter(torch.empty(1))], *args[1:], **kwargs)
+        # It is necessary to make sure that what's holding the parameters is not an iterator, otherwise it can lead to
+        # unsuspected behaviour since each entry will be evaluated at iteration time. There are 2 possibilities:
```
Comment (dacorvo) [inline optimum/neuron/distributed/utils.py:845]:
```suggestion
        # unexpected behaviour since each entry will be evaluated at iteration time. There are 2 possibilities:
```

File: optimum/neuron/utils/cache_utils.py [hunk 2]
```diff
@@ -266,9 +267,13 @@ def get_num_neuron_cores() -> int:
...
-    stdout = stdout.decode("utf-8")
-    json_stdout = json.loads(stdout)
-    return sum(neuron_device_info["nc_count"] for neuron_device_info in json_stdout)
+    if proc.returncode != 0:
+        num_cores = 0
+    else:
+        stdout = stdout.decode("utf-8")
+        json_stdout = json.loads(stdout)
+        num_cores = sum(neuron_device_info["nc_count"] for neuron_device_info in json_stdout)
+    return num_cores
```
Comment (dacorvo) [inline optimum/neuron/utils/cache_utils.py:276]: I think you can also simply count the number of pseudo-devices under /dev, like I did here: https://github.com/huggingface/optimum-neuron/blob/a00d42513765adb4d78eaeabf36fc4b89d4a92eb/optimum/neuron/modeling_decoder.py#L211

**Review by dacorvo — APPROVED** [review]
Just a few nits. LGTM, thanks !

File: tests/distributed/test_model_parallelization.py [hunk 3]
```diff
@@ -141,390 +161,305 @@ def _generate_supported_model_class_names(
...
-            overwrite_model_config=config_overwrite,
+        xla_inputs = {k: v.to(xm.xla_device()) for k, v in inputs.items()}
+        xm.mark_step()
+
+        with torch.no_grad():
+            orig_model_outputs = orig_model(**xla_inputs)
+
+        xm.mark_step()
+
+        # The parallel model needs to be define after the forward pass of the first model because there is a
```
Comment (JingyaHuang) [inline tests/distributed/test_model_parallelization.py:357]:
```suggestion
        # The parallel model needs to be defined after the forward pass of the first model because there is a
```

**Review by JingyaHuang — COMMENTED** [review]
:fire: Awesome work for enabling PP, and improving training tests!! Just left some small nits.

File: tests/conftest.py [hunk 4]
```diff
@@ -70,3 +70,30 @@ def inf_decoder_model(request):
 @pytest.fixture(scope="module", params=[INFERENTIA_MODEL_NAMES[model_arch] for model_arch in DIFFUSER_ARCHITECTURES])
 def inf_diffuser_model(request):
     return request.param
+
+
+# This hook is run before the default pytest_runtest_call
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply tests/conftest.py:75]: Alright, I'm not agaisnt this. The only question I have is: won't it be useful for inference as well?

[the remark this answers is not in the document]
Comment (dacorvo) [reply tests/conftest.py:75]: I don't think so. I usually include setup and teardown directly in the fixtures (this is the recommended way to do it).

File: optimum/neuron/utils/cache_utils.py (hunk 2 above)
Comment (michaelbenayoun) [reply optimum/neuron/utils/cache_utils.py:276]: Nice solution!
Better than what I had done, thanks!

File: examples/image-classification/run_image_classification.py [hunk 5]
```diff
@@ -177,6 +195,15 @@ def main():
     else:
         model_args, data_args, training_args = parser.parse_args_into_dataclasses()
 
+    if model_args.use_auth_token is not None:
+        warnings.warn(
+            "The `use_auth_token` argument is deprecated and will be removed in v4.34. Please use `token` instead.",
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply examples/image-classification/run_image_classification.py:200]: Yes. But I'm not in favor of adding that. These files are updated automatically by cloning the examples from Transformers. This is a bad side effect but I think that's ok.

File: optimum/neuron/accelerate/accelerator.py [hunk 6]
```diff
@@ -434,7 +534,7 @@ def clip_grad_value_(self, parameters, clip_value):
 
     def _custom_save_state(
         self,
-        save_model_func: Callable[["Accelerator", "PreTrainedModel", Union[str, Path], int], Any],
+        save_model_func: Optional[Callable[["Accelerator", "PreTrainedModel", Union[str, Path], int], Any]],
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/neuron/accelerate/accelerator.py:537]: No that's ok, it has to be specified, but the `None` value is allowed.

File: optimum/neuron/accelerate/utils/dataclasses.py [hunk 7]
```diff
@@ -140,21 +140,28 @@ def load_optimizer(self, accelerator, optimizer, model, input_dir, optimizer_ind
...
 
 @dataclass
-class TensorParallelismPlugin:
+class ModelParallelismPlugin:
     tensor_parallel_size: int = 1
     parallelize_embeddings: bool = True
     sequence_parallel_enabled: bool = False
+    pipeline_parallel_size: int = 1
+    pipeline_parallel_num_microbatches: int = 1
+    pipeline_parallel_use_zero1_optimizer: bool = False
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/neuron/accelerate/utils/dataclasses.py:149]: No we can. Everything here can be combined actually!
But you're right. We do have a page in the documentation. I will make sure I update it for PP.

File: tests/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply tests/conftest.py:75]: Alright, will move it then.

**Comment by michaelbenayoun** [comment]
I addressed the comments. @dacorvo @JingyaHuang wdyt?

**Review by dacorvo — APPROVED** [review]
LGTM, thanks !
